src/solver/impls/arkode/arkode_mri.cxx

```diff
@@ -0,0 +1,1090 @@
+
+  // Save derivatives to dudata
+  save_derivs(dudata);
+}
+
+/**************************************************************************
+ * Preconditioner functions
+ **************************************************************************/
+
+void ArkodeMRISolver::pre_s(BoutReal t, BoutReal gamma, BoutReal delta, BoutReal* udata,
+                       BoutReal* rvec, BoutReal* zvec) {
+  TRACE("Running preconditioner: ArkodeMRISolver::pre({:e})", t);
+
+  const BoutReal tstart = bout::globals::mpi->MPI_Wtime();
+
+  if (!hasPreconditioner()) {
+    // Identity (but should never happen)
+    const auto length = N_VGetLocalLength_Parallel(uvec);
+    std::copy(rvec, rvec + length, zvec);
+    return;
+  }
+
+  // Load state from udata (as with res function)
+  load_vars(udata);
+
+  // Load vector to be inverted into F_vars
+  load_derivs(rvec);
+
+  runPreconditioner(t, gamma, delta);
+
+  // Save the solution from F_vars
+  save_derivs(zvec);
+
+  pre_Wtime_s += bout::globals::mpi->MPI_Wtime() - tstart;
+  pre_ncalls_s++;
+}
+
+void ArkodeMRISolver::pre_f(BoutReal t, BoutReal gamma, BoutReal delta, BoutReal* udata,
+                       BoutReal* rvec, BoutReal* zvec) {
+  TRACE("Running preconditioner: ArkodeMRISolver::pre({:e})", t);
+
+  const BoutReal tstart = bout::globals::mpi->MPI_Wtime();
+
+  if (!hasPreconditioner()) {
+    // Identity (but should never happen)
+    const auto length = N_VGetLocalLength_Parallel(uvec);
+    std::copy(rvec, rvec + length, zvec);
+    return;
+  }
+
+  // Load state from udata (as with res function)
+  load_vars(udata);
+
+  // Load vector to be inverted into F_vars
+  load_derivs(rvec);
+
+  runPreconditioner(t, gamma, delta);
+
+  // Save the solution from F_vars
+  save_derivs(zvec);
+
+  pre_Wtime_s += bout::globals::mpi->MPI_Wtime() - tstart;
+  pre_ncalls_s++;
+}
+
+/**************************************************************************
+ * Jacobian-vector multiplication functions
+ **************************************************************************/
+
+void ArkodeMRISolver::jac_s(BoutReal t, BoutReal* ydata, BoutReal* vdata, BoutReal* Jvdata) {
+  TRACE("Running Jacobian: ArkodeMRISolver::jac({:e})", t);
+
+  if (not hasJacobian()) {
+    throw BoutException("No jacobian function supplied!\n");
+  }
+
+  // Load state from ydate
+  load_vars(ydata);
+
+  // Load vector to be multiplied into F_vars
+  load_derivs(vdata);
+
+  // Call function
+  runJacobian(t);
+
+  // Save Jv from vars
+  save_derivs(Jvdata);
+}
+
+void ArkodeMRISolver::jac_f(BoutReal t, BoutReal* ydata, BoutReal* vdata, BoutReal* Jvdata) {
+  TRACE("Running Jacobian: ArkodeMRISolver::jac({:e})", t);
+
+  if (not hasJacobian()) {
+    throw BoutException("No jacobian function supplied!\n");
+  }
+
+  // Load state from ydate
+  load_vars(ydata);
+
+  // Load vector to be multiplied into F_vars
+  load_derivs(vdata);
+
+  // Call function
+  runJacobian(t);
+
+  // Save Jv from vars
+  save_derivs(Jvdata);
+}
```

Comment:
These should be deleted, since the interface functions above never actually attach these functions to ARKODE
```suggestion
```